### ISSUE TITLE
Print layout - label locations in preview different from printed (different at different zooms)

### ISSUE THREAD

**Issue by talflon** [qgis/QGIS#33560]
**Describe the bug**

The locations of labels shown in the Print Layout window, and whether they're visible or not, can be dependent on the zoom level in the window. This is not the zoom scale of the map item, this is the "Zoom level" in the lower-right-hand corner of the screen--the labels move without any changes to the map item, features, or layers. Apparently the label layout algorithm is re-run every time the zoom is changed, and can come to different answers at different zoom levels, despite the map item scale and export resolution being unchanged. As a result, it's not possible for the user to predict what the labels will look like when printed or exported.

**How to Reproduce**

The following file demonstrates the problem:
[qgis-3-labeling-example.zip](https://github.com/qgis/QGIS/files/4010856/qgis-3-labeling-example.zip)

Open the QGZ file, and open the print layout "Layout1". Zoom in and out, without changing the map item scale, and you can see at some zoom levels the labels change: some labels on the polygons disappear, and some labels on the lines jump from one end of their line to the other:

![zoom74 1](https://user-images.githubusercontent.com/9704122/71600554-711d1a00-2b1d-11ea-8859-439bdc2a245a.png)
![zoom37 1](https://user-images.githubusercontent.com/9704122/71600557-75493780-2b1d-11ea-9941-b49e94c4cdf2.png)
![zoom9 3](https://user-images.githubusercontent.com/9704122/71600561-78442800-2b1d-11ea-81f2-04924e9dac9c.png)

The labels on the polygons were set to "Only draw labels which fit completely within feature". The labels on the lines were set "On line". Both were set to sizes in points.

**QGIS and OS versions**

QGIS 

QGIS version
3.10.1-A Coruña
QGIS code revision
ef24c526da
Compiled against Qt
5.11.2
Running against Qt
5.11.2
Compiled against GDAL/OGR
3.0.2
Running against GDAL/OGR
3.0.2
Compiled against GEOS
3.8.0-CAPI-1.13.1
Running against GEOS
3.8.0-CAPI-1.13.1 
Compiled against SQLite
3.29.0
Running against SQLite
3.29.0
PostgreSQL Client Version
11.5
SpatiaLite Version
4.3.0
QWT Version
6.1.3
QScintilla2 Version
2.10.8
Compiled against PROJ
6.2.1
Running against PROJ
Rel. 6.2.1, November 1st, 2019
OS Version
Windows 7 SP 1 (6.1)
Active python plugins
db_manager; 
MetaSearch; 
processing

**Additional context**

While it takes somewhat extreme measures to trigger the problems in this example, this was motivated by another map of mine where the labels were jumping around more easily.

**Comment by roya0045** [comment]
Nyall is reworking the labelling engine and it might affect this behaviour.

**Comment by State-Of-Joshing-Gentle-Peevishness** [comment]
I ran into the same issue in 3.16.3. I assumed that the print-layout would be printed according to the 100% zoom appearance (which I wanted), but the labels of points were in different places when exporting as pdf. Labels of points were set to Mode: Cartographic, Distance: 1,5 mm, Distance offset: From Point: Priority: Highest.
This is really annoying because I don't know which zoom-level I have to choose to preview the print-layout as it will be exported. I solved the issue for me by changing to Mode: Arround Point, which jumped less and gave me a label placement that I could work with.

QGIS version
3.16.3-Hannover
QGIS code revision
94ac9f21b8
Compiled against Qt
5.12.8
Running against Qt
5.12.8
Compiled against GDAL/OGR
3.0.4
Running against GDAL/OGR
3.0.4
Compiled against GEOS
3.8.0-CAPI-1.13.1
Running against GEOS
3.8.0-CAPI-1.13.1 
Compiled against SQLite
3.31.1
Running against SQLite
3.31.1
PostgreSQL Client Version
12.5 (Ubuntu 12.5-0ubuntu0.20.04.1)
SpatiaLite Version
4.3.0a
QWT Version
6.1.4
QScintilla2 Version
2.11.2
Compiled against PROJ
6.3.1
Running against PROJ
Rel. 6.3.1, February 10th, 2020
OS Version
Linux Mint 20.1
Active python plugins
MetaSearch; 
processing

**Comment by roya0045** [comment]
If you export at 96 dpi it shouldn't change much, otherwise 100% won't give you a perfect preview.

**Comment by State-Of-Joshing-Gentle-Peevishness** [comment]
I have to export at 600dpi. I circumvented the issue as described above by changing the mode of label placement.

**Comment by pjshelton** [comment]
This is the same issue as #41476

**Comment by niera87** [comment]
Exactly, it's the same thing. The label locking function is in my opinion very necessary. To avoid this problem, I started using the Easy Custom labeling plugin. However, you cannot use automatic label placement in this situation.

**Comment by pjshelton** [comment]
Is #39490 related to this?

**Comment by Pedro-Murteira** [comment]
Still valid on QGIS 3.22.3. Tested with the provided QGZ file.

**Comment by joeberkovitz** [comment]
I have found that this problem is related to the units employed by the label text. If I set the label font size in terms of Map Units, the problem goes away. If I set the font size in terms of Millimeters, all the labels are offset.

**Comment by JuanBreva** [comment]
Encountered this very issue today using QGIS 3.34.3 running on Mac OS and it is far from be fixed.

**Comment by hachete01** [comment]
I ran into this issue today using QGIS 3.36.1 on Mac OS (ARM). Upon further reading and testing, in my map it seems the issue is caused at higher DPIs (one larger map, exported OK at 200 DPI or less; another smaller map exported OK at 150 DPI or less). It was occurring both for PNG and PDF export, and both when exporting the map canvas view or a print layout. However, in both methods when DPI was set to smaller setting the output was much closer to how the labels looked in the screen.

**Comment by ptrongithub** [comment]
This bug should be labeled as Critical, as it still exists in 2024, QGIS version 3.36.2-Maidenhead, running on Debian 12 Bookworm. Once a layout is rendered, it should not have to be re-rendered when changing zoom, printing, etc. It would also be helpful if there was way to get a better idea of how the labels will be positioned while in the map view (non-layout).

**Comment by saslaw** [comment]
This issue persists in 3.40.5 running on mac OS 15.4.1. Labels appear correctly only in PNG at 96 dpi which is unacceptably low res. At higher resolutions, one label is placed on the opposite side of the polygon from which it appears in layout.
